src/wasm/wat-parser.cpp

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cctype>
+#include <optional>
+#include <variant>
+
+#include "wat-parser.h"
+
+using namespace std::string_view_literals;
+
+namespace wasm::WATParser {
+
+namespace {
+
+// ================
+// Lexical Analysis
+// ================
+
+struct LexCtx {
+protected:
+  std::string_view original;
+  size_t index;
```

Comment:
Is this the start index, or end index? (if the start, is there an end index?)

Reply:
(It's possible "LexCtx" is a known concept that answers these questions, but I'm not familiar it...)

Reply:
This index is the boundary between "lexed" and "next". `lexedSize` would probably be a better name for it.